cdap-app-fabric/src/main/java/io/cdap/cdap/internal/operation/InMemoryOperationRunner.java

```diff
@@ -0,0 +1,156 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(InMemoryOperationRunner.class);
+
+  private final OperationStatePublisher statePublisher;
+  private final Lock lock;
+  private OperationDriver<T> driver;
+
+  /**
+   * Default constructor.
+   *
+   * @param operations A map of {@link OperationType} to {@link LongRunningOperation} injected
+   *     via {@link com.google.inject.multibindings.MapBinder}
+   * @param statePublisher Publishes the current operation state.
+   */
+  @Inject
+  public InMemoryOperationRunner(Map<OperationType, LongRunningOperation> operations,
+      OperationStatePublisher statePublisher) {
+    super(operations);
+    this.statePublisher = statePublisher;
+    this.lock = new ReentrantLock();
+  }
+
+  // TODO(samik) It is currently rudimentary implementation to unblock the state management
+  //  Handle states, publish failures and heartbeat messages
+  @Override
+  public void run(LongRunningOperationRequest<T> request)
+      throws OperationTypeNotSupportedException, InvalidRequestTypeException {
+    lock.lock();
+    try {
+      if (driver != null) {
+        throw new IllegalStateException("Operation already running");
+      }
+      driver = new OperationDriver<>(getOperation(request), request, statePublisher);
+      driver.start();
+      statePublisher.publishRunning();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  // TODO(samik): Implement graceful shutdown
+  @Override
+  public void stop(long timeout, TimeUnit timeoutUnit) throws Exception {
+    lock.lock();
+    try {
+      if (driver == null) {
+        throw new IllegalStateException("No operation is running");
+      }
+      driver.stopAndWait();
+      statePublisher.publishStopped();
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private static final class OperationDriver<T> extends AbstractExecutionThreadService {
+
+    private final LongRunningOperation<T> operation;
+    private final LongRunningOperationRequest<T> request;
+
+    private final OperationStatePublisher statePublisher;
+
+    private ExecutorService executor;
+
+    private OperationDriver(LongRunningOperation<T> operation,
+        LongRunningOperationRequest<T> request,
+        OperationStatePublisher statePublisher) {
+      this.operation = operation;
+      this.request = request;
+      this.statePublisher = statePublisher;
+    }
+
+    @Override
+    protected void run() throws Exception {
+      // TODO(samik) see if we can use Futures.addCallback or addListener
+      try {
+        ListenableFuture<OperationError> error = operation.run(
+            request.getOperationRequest(),
+            statePublisher::publishMetaUpdate);
+        if (error.get() != null) {
+          statePublisher.publishFailed(error.get());
+        } else {
+          statePublisher.publishSuccess();
+        }
+      } catch (InterruptedException e) {
+        // TODO(samik) handle InterruptedException cleanly
```

Comment:
This happens when the `triggerShutdown` is called. You should try to cancel the operation run future.